url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/metainformation/MetaInformationTable.java

```diff
@@ -0,0 +1,222 @@
+			return tableCache.get(name);
+		}
+		
+        int ignore = 0;
+        int attributesCount = 0;
+        int firstPage = 0;
+        List<Attribute> attributes = null;
+        for(Record rec: table) {
+            Map<SecondLevelId, Object> recAtributes = rec.getAttributes();
+            if(ignore != 0) {
+                ignore -= 1;
+                continue;
+            }
+            String currName = (String) recAtributes.get(NAME_ID);
+            if(attributes != null) {
+                attributesCount -= 1;
+                Attribute.DataType type = integersToType(recAtributes);
+                Attribute atr = new Attribute(currName, type);
+                attributes.add(atr);
+                if(attributesCount == 0) {
+            		Map<Attribute, BxTree> indexes = loadIndexes(name, attributes);
+                    Table result = new Table(bufferManager, firstPage, 
+                    		attributes, name, indexes);
+                    if(tableCache.size() == MAX_TABLE_CACHE_ITEMS) {
+                    	tableCache.remove(tableCache.keySet().iterator().next());
+                    }
+                    tableCache.put(name, result);
+                    return result;
+                }
+                continue;
+            }
+            if(!name .equals(currName)) {
+                ignore = (Integer)recAtributes.get(VAL2_ID);
+                continue;
+            }
+            attributes = new ArrayList<>();
+            attributesCount = (Integer)recAtributes.get(VAL2_ID);
+            firstPage = (Integer)recAtributes.get(VAL1_ID);
+        }
+
+        return null;
+    }
+
+    private Map<Attribute, BxTree> loadIndexes(String tableName, List<Attribute> attributes) {
+    	Map<Attribute, BxTree> indexes = new HashMap<>();
+    	for(Record rec: indexesTable) {
+    		if(!tableName.equals((String) rec.getAttributes().get(TABLE_NAME))) {
+    			continue;
+    		}
+    		Integer firstPage = (Integer) rec.getAttributes().get(INDEX_FIRST_PAGE);
+    		Integer atrIdx = (Integer) rec.getAttributes().get(ATR_IDX);
+    		indexes.put(attributes.get(atrIdx), new BxTree(bufferManager, firstPage));
+    	}
+		return indexes;
+	}
+    
+    public boolean createIndex(String tableName, int atributeIdx, String attributeName) {
+    	for(Record rec: indexesTable) {
+    		if(!tableName.equals((String) rec.getAttributes().get(TABLE_NAME))) {
+    			continue;
+    		}
+    		Integer atrIdx = (Integer) rec.getAttributes().get(ATR_IDX);
+    		if(atrIdx == atributeIdx) {
+    			// Index already exists
+    			return false;
+    		}
+    	}
+    	try {
+        	Table table = loadTable(tableName);
+        	int indexFirstPage = bufferManager.getFreePage();
+        	Object[] indexRecord = {tableName, atributeIdx, indexFirstPage};
+			indexesTable.insertRecord(indexRecord);
+			
+			BxTree index = new BxTree(bufferManager, indexFirstPage);
+			SecondLevelId atrId = new SecondLevelId(tableName, attributeName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;attributeName&#39; (index: 2 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;tableName&#39;, &#39;atributeIdx&#39;. The same parameter type is validated here for example: url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/common/DBANTLRErrorListener.java:28. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).